### DISCUSSION TITLE
Phantom Programmable Typing language System - All Types to be back with phantom types, further design time constrains, to communicate the correct way and behaviors of a programs use

### DISCUSSION THREAD

**Discussion by wesleyolis** [dotnet/csharplang#6541] [Language Ideas]
Good Day,

I would like to suggest the following idea, which we got in typescript language, to do somethings their, while hacking the typing system types, to make it do more usefully things. I have subsequently feel and think that any other language at design time would benefit from having the concept of a phantom types engine build into/o ntop of there existing typing engine.
The benefits are two fold, it gives us as the programmer, many more tools and ways, in which we can design and document patterns and behaviors or how we intend out software to be used and for it to be efficiently used correct at design time.

I feel that phantom typing language system would allow for us to better document, better solve and better communicate and reduce mistakes. I see application in many space I have been before, were hacks get me close but not there just there. I see applications in locking validation and ensure of behavior, variables associated with lock scopes, sorting constraints, data set constraints, performance boundaries validation between classes of different layers guarantees (no mistakes) and mitigation measures, async communication behavior(documentation) assurance of hook in points for predefined expect event order behavior, which is a problem with dealing with multiple phone/device vendors APIs monard type event system at design time.

Basically if concrete type in the C# language or any language for that matter, Types had a hidden no programmatical execution type that is pass around with it, just like a monard if I understood that correctly form the 7 page derivation.

What I would like to be able to do is with phantom types is further constrain types and validation function signatures and scoping/blocking. So this could be a new syntax introduce, which allows one to access and manipulate the phantom design time types of concreate language types.

Basically I would say Phantom Typing Language Systems,
Is just key value pair of neast dictionaries variable language and list, and sets, in which, basically contains, hasAtLeastOne, not contains, then also class inheritance concepts for virtuals..

I can't highlight to your all the syntax in short space of time, but the essence of the idea can be expaned on. Think of what Typescript is to Javascript, however, Typescript still needs phantom types, which what this would be. Think of Phantom Typing Language in ways that Typescript is for Javascript, however, with out language changes only programmable typings.

I shall add more examples as I get time, but hopefully this gets you thinking.

```.csharp

function func{Phantom Types constraints}<Generics>(params)

PType someCommonProcedureLock = PScopeType;

class QueueBuffer {
object someCommonProdedureLock = new {scope:default:PTypeScopeObject}Object();
function void {SomeCOmmonProcedureLock} getLock(){
return this.someCommonProdedureLock {some phantom specific casting or return of internal phantom type constraints};
}
 //Just like in Typescript for JavaScript you can do multiple Phantom type signature overloads
 // scope has a default member key of default, if not specific or this or callie, its a convention.
Metrics {scope extends PTypeScopeObject}GetMetrics(int params){
} 

void {SomeCOmmonProcedureLock} getLock(){
return this.someCommonProdedureLock {some phantom specific casting or return of internal phantom type constraints};
}

Metrics {scope[namescope1] extends PTypeScopeObject}GetMetrics2(int params){
} 
}
using (queueBuffer.getLock())
{
// Introduce a phantom type block scope and place it on the stack.

// Failed, Phaton Type requires scope[namescope] to extend PTypeScopeObject
var metrics = QueueBufferInstance.GetMetrics2(1);
}

```
Then also for sorting functions one could communicate and keep a phanton, donating natural sort order of a structure, then use this as a contains for later function calls that are optimized for this.
```csharp
void {dataMeta[sortedKey:1]}Test(int key){
var list = new List<>();
return list {dataMeta[sortedKey:1]};
}
```
Gotta run, type flesh out later more, drafts would be good.

**Comment by YairHalberstadt** [comment]
Typescript adds this to JavaScript as JavaScript is a dynamically typed language, but C# is already strongly typed, so why does it need another type system on top of that?

**Reply by wesleyolis** [reply]
yes it is strongly type, however, we are not able to innovate enough in any language so far that I am a where (c#, C,C++, Typescript, Dephi, Java)  of further with this concept.
By programmable types on top of Strongly type languages, we can dynamically constrain and further constrain and validate the behavior of a program at design compile, binding time. The now strongly type language can be more specific and precise for behaviors and things. Currently I have hacked typescript as far as it can be pushed in this space, its now limited and typescript needs it own Phantom types, which I have suggested many years back say 4 years ago, but there more application for it in other languages too at design time would benefit.  Right now I am working in C# and look at the best way to design certain things and communicate things clearly.
Sadly for a high performance application, such when dealing with locking, make your function name rather explicit and descript of the locks, is the best way. However, in many cases if typescript and other languages had phantom types, I could better convey, and communicate the API behaviors. Sadly I need this in all languages in C++ I have looked at how I could use generics and macros potentially to implement and at design time validate for class performance calls, ensure all these calls everything is in memory, no disk hits.

I have also look at on paper how in typescript for instance could I write async event state life cycle system, were I could provide guaranteed places in the runtime program lifecycle behavior for consumer to hook into the event chain, were the async event ordering and behavior is guaranteed and all, I had a few pages. As basically writing a sync program, with construct, that allows specific the hook in points, however, when it is run and compiled, were the sync program sections are as async and the timing is guaranteed for event order for sync hook point and only the synchronized the behavior of API calls for guarantee of order, but one needs ways to communicate these design constraint, so then question becomes how do I hack typescript typing engine, to make this happen or a POC or do I need build a whole phantom type system myself, which I don't really have the time to do. What I came up with was a way to define all the lifecycles and things like that in a programmatic way struct, however, I then needed to convert that to type constraints and API, so could only be used in intended way at design time it assure to be correct.

One basically then has to write alot less test as all compile time validated and will work with no questions asked, no testing nothing.

Currently, there are ways of doing this, but then basically passing in dummy variables as parameters to functions, to use a strongly types system to validate the phantom types, they have no meaning and basically just there further constraint behaviors and intend use cases.
Basically don't want all these variables in your program, as they just slowing down all function calls, there are limits to what one could do with that approach.

```.csharp

class SpesificLock : Object
{

}

Class QueueBuffer
{
SpesificLock getLock();
}


var lock = queueBuffer.GetLock();

lock(lock)
{
        var metrics = queueBufferInstnace.GetMetrics(lock); // unessary type validation, communicate the behaviors.
        // one doesn't exactly in this case want the lock in the function, because many API class, which can be done with a lock.
        // depending on the design output wanted, so best to batch them up until a single scoped lock.
}
```

Then there is this, which happens so often in the deep call stacks of programs, were miscall and have performance bounder violation.

```csharp
enum PerfConClass
{
OnDie = 100,
L1 = 200,
L2 = 300,
L3 = 400,
RAM1 = 500
SSD = 600
Disk = 700
TAP = 800
}

class
{
{ClassPerformanceDesignConstraintRequirement}: PerfConClass.L1// define phantom types, all initialize variables will have to have this contains in the type format.
{this.ClassCallingFunctionDesignConstraint}: PerfConClass.InMem; // implicit phantom type.

void ensurefunctionCallConstrain{extends ClassPerformanceDesignConstraintRequirement}(

otherclass.callotherclass()// Implicit call of validation
}
// one would need global class extends for function definitions of classes for this.
void callotherclass{extends PerfConClass.InMem}()
}
```

**Comment by HaloFour** [comment]
I might be missing something here but from my understanding of phantom types they are something already trivially supported by generics in C# and the .NET runtime?  I don't know if Typescript makes use of them in any specific manner or with some specific syntax that enables more scenarios.  As mentioned, Typescript has a lot of flexibility in how it's type system works given that it's completely elided when transpiled to Javascript.

**Comment by wesleyolis** [comment]
The main thing is that strong typed language provides some form of constraints for data, however, to full constraint a program to its intended assemble of Lego block parts, with generics and strong types, one is required to further limit and constrain the strong types on another 3 dimensions, by given them some form of phantom type more specific definition type definition, based on the software/program one is writing, which would allow one to further constrain strong type language types by specializing them with phantom types, to further constraint behaviors /event ordering of programs as well how lego bocks would fit to gather for guarantees, that at design time things would just work, with out having to write tests or further more any documentation, as it would allow for more self documentation. Or have to think of the behaviors to test, as the person who wrote the API, has already limited that, to the valid options, there can be no ambiguity, uncertainty,  interpretation issues anymore, because the doc are not good enough at covering all the possible permutations, in which it leaves wholes, which require testing and validating the API. behavior. Think of lawyers, they read between the lines of the documents, for what contains ambiguity, we can't have that, which means basically have to write docs at the level of a law then.
In the phantom types, one could pass alot more background information to do more complex thing than what strong type languages allow. It would allow dynamic design time checking of programs, were many possibilities of assembly, allow them to be even further limited, by phantom types, were currently, one would have to read the manual and they have to provide testing frame works.
That validation, could be converted into programmable phantom types. Cutting the writing of many testing and testing of frameworks, as at design time they would be validated. This would allow for vendors and specialize of complex components to write more correct implementation of things much faster.

C# Generics in falls long way short of typescript, I find myself writing generic class, however, due to generics limitations,
you can't write complete generic complex classes that one requires.

One needs to be able to define strong types bases on generics as type definition for generic class instance as well, this short coming of typescript, few hacks to work around this to return type definition from generic specialization of a  class.

Example:

````.csharp
class ExampleA<TA, TB, TC, TD>
{
TYPE ExampleBSpecializedMore<TA,TB,TC, TD = Enum.Y> // there hacks to do this in typescript, I thought of, which lowers the mental load, as use simplified type names.
}

var temp = new ExampleA<A,B,C ,D>();

var test2 = new temp.ExampleBSpecializedMore() // typescript typeof temp.ExampleBSpecializedMore() where one like hacks the return type to do this.

```
